# SQL FUNCTIONS


Here’s an [ER (Entity Relationship) diagram](https://www.lucidchart.com/pages/er-diagrams) of the chinook database:

![Chinook ERD](https://github.com/Explore-AI/Pictures/blob/master/sqlite-sample-database-color.jpg?raw=true)

_[Image source](https://www.sqlitetutorial.net/sqlite-sample-database/)_

## Important
#### Run these commands to get SQL working in jupyter lab

In [2]:
#load sql magics
%load_ext sql

In [3]:
#load chinook database
%sql sqlite:///chinook.db

'Connected: @chinook.db'

## Basic Sql 

In [26]:
%%sql

select * 
from customers
limit 5;

 * sqlite:///chinook.db
Done.


[(1, 'Luís', 'Gonçalves', 'Embraer - Empresa Brasileira de Aeronáutica S.A.', 'Av. Brigadeiro Faria Lima, 2170', 'São José dos Campos', 'SP', 'Brazil', '12227-000', '+55 (12) 3923-5555', '+55 (12) 3923-5566', 'luisg@embraer.com.br', 3),
 (2, 'Leonie', 'Köhler', None, 'Theodor-Heuss-Straße 34', 'Stuttgart', None, 'Germany', '70174', '+49 0711 2842222', None, 'leonekohler@surfeu.de', 5),
 (3, 'François', 'Tremblay', None, '1498 rue Bélanger', 'Montréal', 'QC', 'Canada', 'H2G 1A7', '+1 (514) 721-4711', None, 'ftremblay@gmail.com', 3),
 (4, 'Bjørn', 'Hansen', None, 'Ullevålsveien 14', 'Oslo', None, 'Norway', '0171', '+47 22 44 22 22', None, 'bjorn.hansen@yahoo.no', 4),
 (5, 'František', 'Wichterlová', 'JetBrains s.r.o.', 'Klanova 9/506', 'Prague', None, 'Czech Republic', '14700', '+420 2 4172 5555', '+420 2 4172 5555', 'frantisekw@jetbrains.com', 4)]

In [35]:
%%sql
select * 
from tracks
limit 5

 * sqlite:///chinook.db
Done.


[(1, 'For Those About To Rock (We Salute You)', 1, 1, 1, 'Angus Young, Malcolm Young, Brian Johnson', 343719, 11170334, 0.99),
 (2, 'Balls to the Wall', 2, 2, 1, None, 342562, 5510424, 0.99),
 (3, 'Fast As a Shark', 3, 2, 1, 'F. Baltes, S. Kaufman, U. Dirkscneider & W. Hoffman', 230619, 3990994, 0.99),
 (4, 'Restless and Wild', 3, 2, 1, 'F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. Dirkscneider & W. Hoffman', 252051, 4331779, 0.99),
 (5, 'Princess of the Dawn', 3, 2, 1, 'Deaffy & R.A. Smith-Diesel', 375418, 6290521, 0.99)]

## Distinct
 - returns all unique values from a coloumn
 - usefull for knowing which artists are represented in a table

In [9]:
%%sql

SELECT DISTINCT composer
FROM tracks 
limit 5

 * sqlite:///chinook.db
Done.


[('Angus Young, Malcolm Young, Brian Johnson',),
 (None,),
 ('F. Baltes, S. Kaufman, U. Dirkscneider & W. Hoffman',),
 ('F. Baltes, R.A. Smith-Diesel, S. Kaufman, U. Dirkscneider & W. Hoffman',),
 ('Deaffy & R.A. Smith-Diesel',)]

## Count
- Counts the number of entries in a coloumn

In [10]:
%%sql

SELECT COUNT(composer)
from tracks


 * sqlite:///chinook.db
Done.


[(2525,)]

## WHERE
- filter with both text and numeric
- uses comparison opperators such as (= ,<,>)
- for multiple "where" use the "AND" opperator

In [14]:
%%sql

SELECT name
from tracks
where composer = 'Angus Young, Malcolm Young, Brian Johnson'
limit 5

 * sqlite:///chinook.db
Done.


[('For Those About To Rock (We Salute You)',),
 ('Put The Finger On You',),
 ("Let's Get It Up",),
 ('Inject The Venom',),
 ('Snowballed',)]

## WHERE/OR
- use the "OR" opperator when the query only requires one of the conditions to be met


In [18]:
%%sql
select composer
from tracks
where name = 'Put The Finger On You' or name = 'Fast As A Shark'

 * sqlite:///chinook.db
Done.


[('Angus Young, Malcolm Young, Brian Johnson',)]

## BETWEEN
- use between to select the range between the conditions

In [20]:
%%sql
select name
from tracks
where milliseconds between 250000 and 300000
limit 5

 * sqlite:///chinook.db
Done.


[('Restless and Wild',),
 ('Evil Walks',),
 ('Breaking The Rules',),
 ('Spellbound',),
 ('Bad Boy Boogie',)]

## WHERE IN
- where in allows us to quickly assign multiple OR conditions

In [27]:
%%sql

select firstname , lastname
from customers
where state in ('SP' , 'QC')

 * sqlite:///chinook.db
Done.


[('Luís', 'Gonçalves'),
 ('François', 'Tremblay'),
 ('Eduardo', 'Martins'),
 ('Alexandre', 'Rocha')]

## IS NULL
- null is a missing or unknown value in sequel
- can combine with "WHERE" for usefull results

In [29]:
%%sql

SELECT firstname , lastname
FROM customers
WHERE fax IS NULL
limit 5

 * sqlite:///chinook.db
Done.


[('Leonie', 'Köhler'),
 ('François', 'Tremblay'),
 ('Bjørn', 'Hansen'),
 ('Helena', 'Holý'),
 ('Astrid', 'Gruber')]

## LIKE/NOT LIKE
- LIKE can be used in a WHERE clause to search for a pattern
- we use the % Wildcard as a place holder for 0 one or many chars


In [32]:
%%sql

select lastname
from customers
where firstname like 'B%'
limit 5

 * sqlite:///chinook.db
Done.


[('Hansen',)]

- _ Wildcard will match only a single char
- this can be usefull for finding specific details
- this example will only return emaails that have 3 chars after the .

In [34]:
%%sql

select firstname , lastname , email
from customers
where email like '%.___'
limit 5

 * sqlite:///chinook.db
Done.


[('François', 'Tremblay', 'ftremblay@gmail.com'),
 ('František', 'Wichterlová', 'frantisekw@jetbrains.com'),
 ('Helena', 'Holý', 'hholy@gmail.com'),
 ('Frank', 'Harris', 'fharris@google.com'),
 ('Jack', 'Smith', 'jacksmith@microsoft.com')]

## AGERGATE FUNCTION
 - avg (calculates the average)
 - max (returns the maximum value)
 - min (returns the minimum value)
 - sum (returns total of adding up all cells)

In [36]:
%%sql
Select avg(milliseconds)
from tracks

 * sqlite:///chinook.db
Done.


[(393599.2121039109,)]

In [37]:
%%sql
select sum(milliseconds)
from tracks

 * sqlite:///chinook.db
Done.


[(1378778040,)]

In [39]:
%%sql
select min(milliseconds)
from tracks

 * sqlite:///chinook.db
Done.


[(1071,)]

In [40]:
%%sql
select max(milliseconds)
from tracks

 * sqlite:///chinook.db
Done.


[(5286953,)]

## ALIASING ---- AS
- use AS to assign a shorthand to a cell

In [41]:
%%sql
select firstname AS n , Lastname AS l
from customers
limit 5

 * sqlite:///chinook.db
Done.


[('Luís', 'Gonçalves'),
 ('Leonie', 'Köhler'),
 ('François', 'Tremblay'),
 ('Bjørn', 'Hansen'),
 ('František', 'Wichterlová')]

## ORDER BY
- orders the data in asceding order 
- can reverse order with DESC 

In [48]:
%%sql
select city , firstname , lastname
from customers
order by city desc
limit 10

 * sqlite:///chinook.db
Done.


[('Yellowknife', 'Ellie', 'Sullivan'),
 ('Winnipeg', 'Aaron', 'Mitchell'),
 ('Warsaw', 'Stanisław', 'Wójcik'),
 ('Vienne', 'Astrid', 'Gruber'),
 ('Vancouver', 'Jennifer', 'Peterson'),
 ('Tucson', 'Patrick', 'Gray'),
 ('Toronto', 'Robert', 'Brown'),
 ('São Paulo', 'Eduardo', 'Martins'),
 ('São Paulo', 'Alexandre', 'Rocha'),
 ('São José dos Campos', 'Luís', 'Gonçalves')]

- can sort multiple cloums at a time BUT it will sort in the order that you have queried it

In [49]:
%%sql
select city , firstname , lastname
from customers
order by city desc , firstname 
limit 10


 * sqlite:///chinook.db
Done.


[('Yellowknife', 'Ellie', 'Sullivan'),
 ('Winnipeg', 'Aaron', 'Mitchell'),
 ('Warsaw', 'Stanisław', 'Wójcik'),
 ('Vienne', 'Astrid', 'Gruber'),
 ('Vancouver', 'Jennifer', 'Peterson'),
 ('Tucson', 'Patrick', 'Gray'),
 ('Toronto', 'Robert', 'Brown'),
 ('São Paulo', 'Alexandre', 'Rocha'),
 ('São Paulo', 'Eduardo', 'Martins'),
 ('São José dos Campos', 'Luís', 'Gonçalves')]

- see how Alexandre and Eduardo have switch postions

## GROUP BY 
 - allows sorting of data
 - commonly used with AGGREGATE functions
 - always goes after the FROM statement

In [53]:
%%sql
select state, count(*)
from customers
group by state
limit 5

 * sqlite:///chinook.db
Done.


[(None, 29), ('AB', 1), ('AZ', 1), ('BC', 1), ('CA', 3)]

## HAVING
 - AGGREGATE functions CANNOT be used in a WHERE clause
 - Filtering with an AGGREGATE function in done by using HAVING

In [65]:
%%sql
select  milliseconds
from tracks
group by milliseconds
having count(name) > 3

 * sqlite:///chinook.db
Done.


[(240091,)]

- this code shows us that there are 4 or more songs that are 240091 milliseconds long

## Creating a Table
CREATE TABLE table_name ( \
    column_name data type \
    );

In [39]:
%%sql
create table test2(
    firstname text,
    lastname text,
    job text
);

select * 
from test

 * sqlite:///chinook.db
Done.
Done.


[]

## Altering a Table
ALTER TABLE table_name
    ADD COLUMN column_name data_type

In [40]:
%%sql
alter table test2
add location text ;
    
    
select * 
from test2

 * sqlite:///chinook.db
Done.
Done.


[]

## Rename & Drop
- Renaming a column \
 ALTER TABLE table_name\
RENAME COLUMN old_name TO new_nam

In [41]:
%%sql
alter table test2
rename column job to occupation;

select * 
from test2;


 * sqlite:///chinook.db
Done.
Done.


[]

- Dropping a column \
ALTER TABLE table_name \
DROP COLUMN column_name;


In [43]:
%%sql
alter table test2
drop column location;

alter table test2
drop column occupation;

select *
from test2


 * sqlite:///chinook.db
Done.
Done.
Done.


[]

## Insert Into
INSERT INTO table_name \
SELECT DISTINCT Column_name_a , column_name_b \
FROM table_name ;

In [44]:
%%sql 
insert into test2
select distinct firstname , lastname
from customers;

select *
from test2
limit 5;

 * sqlite:///chinook.db
59 rows affected.
Done.


[('Luís', 'Gonçalves'),
 ('Leonie', 'Köhler'),
 ('François', 'Tremblay'),
 ('Bjørn', 'Hansen'),
 ('František', 'Wichterlová')]